### PR TITLE
[Redesign] Update ObstacleLocationFinderOperator, FusionOperator

### PR DESCRIPTION
**Pull request by objorkman**
* Also, updates to `operator_creator` functions for obstacle tracking

### CODE REVIEW

File: pylot/component_creator.py [hunk 1]
```diff
@@ -70,10 +70,10 @@ def add_obstacle_detection(center_camera_stream,
         if any('efficientdet' in model
                for model in FLAGS.obstacle_detection_model_names):
             logger.debug('Using EfficientDet obstacle detector...')
-            obstacles_streams = pylot.operator_creator.\
+            obstacles_stream = pylot.operator_creator.\
```
Comment (pschafhalter) [inline pylot/component_creator.py:73]: Could you add type annotations to `def add_obstacle_detection(center_camera_stream: erdos.Stream, ...) -> Tuple[erdos.Stream, erdos.Stream]`?

File: pylot/operator_creator.py [hunk 2]
```diff
@@ -33,17 +33,17 @@ def add_efficientdet_obstacle_detection(camera_stream,
         EfficientDetOperator
     if csv_file_name is None:
         csv_file_name = FLAGS.csv_log_file_name
-    op_config = erdos.OperatorConfig(name='efficientdet_operator',
```
Comment (pschafhalter) [inline pylot/operator_creator.py:36]: Please add type annotations to `def add_efficientdet_obstacle_detection(`

File: pylot/operator_creator.py [hunk 3]
```diff
@@ -33,17 +33,17 @@ def add_efficientdet_obstacle_detection(camera_stream,
...
+        csv_log_file_name=csv_file_name,
+        profile_file_name=FLAGS.profile_file_name)
+    obstacles_stream = erdos.connect_two_in_one_out(
+        EfficientDetOperator, op_config, camera_stream,
+        time_to_decision_stream, FLAGS.obstacle_detection_model_names,
+        FLAGS.obstacle_detection_model_paths, FLAGS)
+    return obstacles_stream
 
 
 def add_obstacle_detection(camera_stream,
```
Comment (pschafhalter) [inline pylot/operator_creator.py:49]: Please add type annotations

File: pylot/operator_creator.py [hunk 4]
```diff
@@ -90,15 +91,17 @@ def add_obstacle_location_finder(obstacles_stream, depth_stream, pose_stream,
     """
     from pylot.perception.detection.obstacle_location_finder_operator import \
```
Comment (pschafhalter) [inline pylot/operator_creator.py:92]: Please add type annotations to `def add_obstacle_location_finder(...)`

File: pylot/operator_creator.py [hunk 5]
```diff
@@ -610,17 +613,20 @@ def add_localization(imu_stream,
 def add_fusion(pose_stream, obstacles_stream, depth_stream,
```
Comment (pschafhalter) [inline pylot/operator_creator.py:613]: Please add type annotations

File: pylot/perception/detection/obstacle_location_finder_operator.py [hunk 6]
```diff
@@ -17,34 +22,14 @@ class ObstacleLocationFinderOperator(erdos.Operator):
...
-                 depth_stream: erdos.ReadStream, pose_stream: erdos.ReadStream,
-                 obstacles_output_stream: erdos.WriteStream, flags,
-                 camera_setup):
-        obstacles_stream.add_callback(self.on_obstacles_update)
-        depth_stream.add_callback(self.on_depth_update)
-        pose_stream.add_callback(self.on_pose_update)
-        erdos.add_watermark_callback(
-            [obstacles_stream, depth_stream, pose_stream],
-            [obstacles_output_stream], self.on_watermark)
+    def __init__(self, flags, camera_setup):
```
Comment (pschafhalter) [inline pylot/perception/detection/obstacle_location_finder_operator.py:32]:
```suggestion
    def __init__(self, flags, camera_setup: pylot.drivers.sensor_setup.CameraSetup):
```

File: pylot/perception/detection/obstacle_location_finder_operator.py [hunk 7]
```diff
@@ -54,46 +39,51 @@ def __init__(self, obstacles_stream: erdos.ReadStream,
...
+        elif isinstance(data, DepthFrame):
+            self.on_depth_update(context, data)
+        elif isinstance(data, pylot.utils.Pose):
+            self.on_pose_update(context, data)
+        else:
+            raise ValueError('Unexpected data type')
 
-    def destroy(self):
-        self._logger.warn('destroying {}'.format(self.config.name))
+    def on_obstacles_update(self, context, data):
```
Comment (pschafhalter) [inline pylot/perception/detection/obstacle_location_finder_operator.py:54]:
```suggestion
    def on_obstacles_update(self, context: OneInOneOutContext, data: ObstaclesMessageTuple):
```

File: pylot/perception/detection/obstacle_location_finder_operator.py [hunk 8]
```diff
@@ -54,46 +39,51 @@ def __init__(self, obstacles_stream: erdos.ReadStream,
...
+        self._logger.debug('@{}: obstacles update'.format(context.timestamp))
+        self._obstacles_msgs.append(data)
+
+    def on_depth_update(self, context, data):
```
Comment (pschafhalter) [inline pylot/perception/detection/obstacle_location_finder_operator.py:58]:
```suggestion
    def on_depth_update(self, context: OneInOneOutContext, data: DepthFrame):
```

File: pylot/perception/detection/obstacle_location_finder_operator.py [hunk 9]
```diff
@@ -54,46 +39,51 @@ def __init__(self, obstacles_stream: erdos.ReadStream,
...
+        self._logger.debug('@{}: depth update'.format(context.timestamp))
+        self._depth_msgs.append(data)
+
+    def on_pose_update(self, context, data):
```
Comment (pschafhalter) [inline pylot/perception/detection/obstacle_location_finder_operator.py:62]:
```suggestion
    def on_pose_update(self, context: OneInOneOutContext, data: pylot.utils.Pose):
```

File: pylot/perception/fusion/fusion_operator.py [hunk 10]
```diff
@@ -17,21 +22,7 @@ class FusionOperator(erdos.Operator):
...
-                 camera_fov=np.pi / 4,
-                 rgbd_max_range=1000):
-        self.pose_stream = pose_stream
-        self.obstacles_stream = obstacles_stream
-        self.depth_camera_stream = depth_camera_stream
-        pose_stream.add_callback(self.update_pos)
-        obstacles_stream.add_callback(self.update_obstacles)
-        depth_camera_stream.add_callback(self.update_distances)
-        self._fused_stream = fused_stream
+    def __init__(self, flags, camera_fov=np.pi / 4, rgbd_max_range=1000):
```
Comment (pschafhalter) [inline pylot/perception/fusion/fusion_operator.py:25]:
```suggestion
    def __init__(self, flags, camera_fov: float = np.pi / 4, rgbd_max_range: float = 1000):
```

File: pylot/perception/fusion/fusion_operator.py [hunk 11]
```diff
@@ -99,22 +97,19 @@ def fuse(self):
...
+
+    def update_pos(self, context, msg):
+        vehicle_pos = ((msg.transform.location.x, msg.transform.location.y,
+                        msg.transform.location.z),
+                       (msg.transform.forward_vector.x,
+                        msg.transform.forward_vector.y,
+                        msg.transform.forward_vector.z))
+        self._car_positions.append((context.timestamp, vehicle_pos))
+
+    def update_obstacles(self, context, msg):
```
Comment (pschafhalter) [inline pylot/perception/fusion/fusion_operator.py:112]:
```suggestion
    def update_obstacles(self, context: OneInOneOutContext, obstacles: ObstaclesMessageTuple):
```

File: pylot/perception/fusion/fusion_operator.py [hunk 12]
```diff
@@ -123,17 +118,7 @@ def update_obstacles(self, msg):
...
-            pose_msg = self.pose_stream.read()
-            obstacles_msg = self.obstacles_stream.read()
-            depth_camera_msg = self.depth_camera_stream.read()
-            self.update_pos(pose_msg)
-            self.update_obstacles(obstacles_msg)
-            self.update_distances(depth_camera_msg)
-            self.fuse()
+        self._obstacles.append((context.timestamp, vehicle_bounds))
+
+    def update_distances(self, context, msg):
```
Comment (pschafhalter) [inline pylot/perception/fusion/fusion_operator.py:123]:
```suggestion
    def update_distances(self, context: OneInOneOutContext, depth_frame: DepthFrame):
```

**Comment by objorkman** [comment]
Thanks for the comments, latest commit should've added all the type annotations!

**Review by pschafhalter — APPROVED** [review]
LGTM

**Pull request merged**